pkg/datapath/linux/routing/migrate.go

```diff
@@ -0,0 +1,628 @@
+	// either the new state or the old state, and want to avoid being
+	// in-between datapath states.
+	//   1) Copy over new routes from the old routes
+	//   2) Insert new rule
+	//   3) Delete old rule
+	//   4) Delete old routes
+	// Doing (1) & (2) before (3) & (4) allows us to essentially perform an
+	// "atomic" swap-in for the new state.
+	//
+	// If a failure occurs at (1), then the old state can continue to service
+	// the endpoint. Similarly with (2) because routes without rules are likely
+	// to not have any effect.
+	//
+	// If a failure occurs, (3) or (4), then we already have the new state
+	// in-place to direct traffic for the endpoint. There is a slight nuance
+	// here; the older rule has a _higher_ priority which means it would be
+	// selected by Linux first, before reaching the new rule. Therefore, if we
+	// fail at (3), keep in mind we do have the new state in-place, but it goes
+	// "untouched". We haven't deleted the routes and, therefore the old state
+	// is still in-place as well, servicing the endpoint and allowing no
+	// connectivity disruptions. This is the main reason why the new priority
+	// was chosen to have a lower priority than the original priority value.
+	//
+	// If we fail at (4), then the old rule will have
+	// been deleted and the new state is in-place, which would be servicing the
+	// endpoint. The old routes would just be leftover state to be cleaned up
+	// at a later point.
+	//
+	// It is also important to note that we only revert what we've done on a
+	// per-rule basis if we fail at (2) or (3). This is by design because we
+	// want to ensure that each iteration of the loop is atomic to each
+	// endpoint. Meaning, either the endpoint ends up with the new datapath or
+	// the original.
+
+	// Number of rules (endpoints) successfully migrated and how many failed.
+	var migrated, failed int
+
+	if isUpgrade {
+		for _, r := range v1Rules {
+			// Let's say we have an ENI device attached to the node with
+			// ifindex 3 and interface number 2. The following rule will exist
+			// on the node _before_ migration.
+			//   110:    from 192.168.11.171 to 192.168.0.0/16 lookup 3
+			// After the migration, this rule will become:
+			//   111:    from 192.168.11.171 to 192.168.0.0/16 lookup 12
+			// The priority has been updated to 111 and the table ID is 12
+			// because the interface number is 2 plus the routing table offset
+			// (linux_defaults.RouteTableInterfacesOffset). See copyRoutes()
+			// for what happens with routes.
+
+			scopedLog := log.WithField("rule", r)
+
+			routes, err := m.rpdb.RouteListFiltered(netlink.FAMILY_V4, &netlink.Route{
+				Table: r.Table,
+			}, netlink.RT_FILTER_TABLE)
+			if err != nil {
+				scopedLog.WithError(err).
+					Error("Failed to migrate endpoint to new ENI datapath due to failure in listing routes associated with rule. " +
+						"Previous datapath is still intact and endpoint connectivity is not affected.")
+
```

Comment:
If there was a separate `m.upgradeRule()` function to handle the inner logic of the loop, many of these error conditions could simply return a `fmt.Errorf(...)` with the unique bits of these error messages, and we would need only a few lines that actually do the logging. I think this would be easier to read since so many of the lines of this function are similar-looking error messages.